unity-client/Assets/Rendering/Culling/CullingController.cs

```diff
@@ -0,0 +1,551 @@
+
+        public UniversalRenderPipelineAsset urpAsset;
+
+        internal ICullingObjectsTracker objectsTracker;
+        private Coroutine updateCoroutine;
+        private float timeBudgetCount = 0;
+        private Vector3 lastPlayerPos;
+        private bool resetObjectsNextFrame = false;
+
+        public delegate void DataReport(int rendererCount, int hiddenRendererCount, int hiddenShadowCount);
+
+        public event DataReport OnDataReport;
+
+
+        public static CullingController Create()
+        {
+            return new CullingController(
+                GraphicsSettings.currentRenderPipeline as UniversalRenderPipelineAsset,
+                new CullingControllerSettings()
+            );
+        }
+
+        public CullingController(UniversalRenderPipelineAsset urpAsset, CullingControllerSettings settings, ICullingObjectsTracker cullingObjectsTracker = null)
+        {
+            if (cullingObjectsTracker == null)
+                objectsTracker = new CullingObjectsTracker();
+            else
+                objectsTracker = cullingObjectsTracker;
+
+            this.urpAsset = urpAsset;
+            this.settings = settings;
+        }
+
+        /// <summary>
+        /// Starts culling update coroutine.
+        /// The coroutine will keep running until Stop() is called or this class is disposed.
+        /// </summary>
+        public void Start()
+        {
+            if (updateCoroutine != null)
+                return;
+
+            updateCoroutine = CoroutineStarter.Start(UpdateCoroutine());
+        }
+
+        /// <summary>
+        /// Stops culling update coroutine.
+        /// </summary>
+        public void Stop()
+        {
+            if (updateCoroutine == null)
+                return;
+
+            CoroutineStarter.Stop(updateCoroutine);
+            updateCoroutine = null;
+        }
+
+        /// <summary>
+        /// Process all sceneObject renderers with the parameters set by the given profile.
+        /// 
+        /// If profile matches the skinned renderer profile in settings, the skinned renderers are going to be used.
+        /// </summary>
+        /// <param name="profile">any CullingControllerProfile</param>
+        /// <returns>IEnumerator to be yielded.</returns>
+        internal IEnumerator ProcessProfile(CullingControllerProfile profile)
+        {
+            Renderer[] renderers = null;
+
+            if (profile == settings.rendererProfile)
+                renderers = objectsTracker.GetRenderers();
+            else
+                renderers = objectsTracker.GetSkinnedRenderers();
+
+            for (var i = 0; i < renderers.Length; i++)
+            {
+                if (timeBudgetCount > maxTimeBudget)
+                {
+                    timeBudgetCount = 0;
+                    yield return null;
+                }
+
+                Renderer r = renderers[i];
+
+                if (r == null)
+                    continue;
```

Comment:
We could also skip renderers that aren't inside the frustrum culling area using the `Renderer.isVisible` property, what do you think?

Reply:
Few issues to go for this approach: 

* `Renderer.isVisible` uses bounding box culling. This wouldn't work with skinned meshes because the animations tend to go off the bounds. 

* If we have out of frustum renderers and they are not processed, we have to mark them as dirty as they go into the frustum to process them again. Ideally, we must do this without frame delay and with a O(1) operation when setting the dirtyness. This kind of lazy updating logic tends to be buggy and usually requires a lot of test coverage.

So, going for this one will be a bit costly. At this stage, I'd prefer to go with the current approach and try new optimizations later.